app/views/admin/privileged_reactions/index.html.erb

```diff
@@ -30,7 +30,7 @@
     </tr>
   </thead>
   <tbody class="crayons-card">
-    <% @privileged_reactions.each do |reaction| %>
+    <% @privileged_reactions.select(&:reactable).each do |reaction| %>
```

Comment:
all of the "Content" possibilities link to an object via the `reactable` - if that's null this is unsafe to link to, filter that out here.

Reply:
I want to say up front that I don't have full context here, so take everything that follows with a grain of salt:

Looking at the code in the `Reactable` module it seems like reactions should be removed when the `reactable` itself gets destroyed:

```ruby
has_many :reactions, as: :reactable, inverse_of: :reactable, dependent: :destroy
```

Could this be a case of older data not properly being cleaned up? Is there a chance we're better off fixing the data, adding an FK constraint, and making sure everything works as intended?

> I tried to accomplish this in the controller but it seems the paginate helper requires a relation rather than an array.

I'm sure we could come with a query producing a relation for this.

Reply:
This is true for Article and Comment reactions - but for User reaction, for `User`, the `has_many :reactions` is by user_id, not reactable_id, and it's a different relation (user generated reactions, rather than reactions to the user by others) and those are not being cleared by callbacks (I opened a separate issue for this, we probably want to add that to the `Users::Delete` service or one of its collaborators.)

Reply:
```sql
select reactable_id as reacted, reactions.user_id as reactor, reactions.id as reaction_id, reactions.created_at from 
reactions 
left outer join users on users.id = reactions.reactable_id
where reactable_type = 'User' 
and category='vomit'
and users.id IS NULL
order by reaction_id DESC;
```

I get 23 of these dangling user references, created between March 2020 and August 2021, and their presence complicates assumptions in the views that the reaction has a valid reactable.

I'll take a stab at adding the reaction cleanup to the user deletion service, and probably a follow up data update to cleanup.

I don't think I understand how to properly add a foreign key to the reactions table that handles the polymorphism (since the "type" is a column, you'd need to either reference the articles table, the comments table, or the users table, based on the (downcased, pluralized) reactable type. If this is a well worn path of db art I'm happy to learn it,  but it didnt' seem obvious to me how.

Reply:
> I'm sure we could come with a query producing a relation for this. 

I believe that's true. I assume it looks very similar to the `users.id IS NOT NULL` left join I used to filter them. I'm happy to retool that (I very much prefer to move this logic into the controller and out of the view template).

Reply:
> This is true for Article and Comment reactions - but for User reaction, for User, the has_many :reactions is by user_id, not reactable_id, and it's a different relation (user generated reactions, rather than reactions to the user by others)

See, that's the context I said I'm missing, thanks for clarifying!

> I'll take a stab at adding the reaction cleanup to the user deletion service, and probably a follow up data update to cleanup.

🎉 This sounds great.

> I don't think I understand how to properly add a foreign key to the reactions table that handles the polymorphism

There are DB patterns for this, they generally involved an additional table (e.g. `reactables`). Again, I came into this review without full context and just naively through some ideas around, we don't need to pursue this line of thought further.

> I'm happy to retool that (I very much prefer to move this logic into the controller and out of the view template).

If you are willing to give this a try I'd be very much in favor of it, this really doesn't seem like something that should happen in a view template.